system/src/system_control_internal.cpp

```diff
@@ -81,30 +83,38 @@ SystemControl g_systemControl;
 SystemControl::SystemControl() :
 #ifdef USB_VENDOR_REQUEST_ENABLE
         usbChannel_(this),
-#endif
-#if HAL_PLATFORM_BLE
-        bleChannel_(this),
 #endif
         appReqHandler_(nullptr) {
 }
 
 int SystemControl::init() {
 #if HAL_PLATFORM_BLE
-    const int ret = bleChannel_.init();
-    if (ret != 0) {
-        return ret;
+    // BleControlRequestChannel::init() should be called only once
+    // as characteristic UUIDs are currently called only once. If BLE prov mode
+    // is enabled, we need to make sure to call characteristic UUIDs in prov
+    // mode system calls for the first time instead of here.
+    // See system_ble_prov_mode()
+    if (!HAL_Feature_Get(FEATURE_DISABLE_LISTENING_MODE)) {
+        const int ret = BleControlRequestChannel::instance(this)->init();
+        if (ret != 0) {
+            return ret;
+        }
     }
 #endif
     return 0;
 }
 
 void SystemControl::run() {
 #if HAL_PLATFORM_BLE
-    bleChannel_.run();
+    BleControlRequestChannel::instance(this)->run();
 #endif
 }
 
 void SystemControl::processRequest(ctrl_request* req, ControlRequestChannel* /* channel */) {
+    if (std::find(particle::system::SystemControl::vecCtrlReq.begin(), particle::system::SystemControl::vecCtrlReq.end(), req->type) != particle::system::SystemControl::vecCtrlReq.end()) {
```

Comment:
Would we like to make this usable only during prov mode?

Reply:
`vecCtrlReq` is a member variable of `SystemControl`, you don't need to prepend the namespace. Besides, `std::find()` cannot guarantee that whether the given type is found or not when the last element is equal to the given type. Let's do it in this way: 
```
ctrl_request_type type = CTRL_REQUEST_INVALID;
for (auto t : vecCtrlReq) {
    if (t == req->type) {
        type = req->type
        break;
    }
}
if (type == CTRL_REQUEST_INVALID) {
    return;
}
```